File: cpp/questions/Trie.cpp
#include <bits/stdc++.h>

using namespace std;

class Trie {
private:
    struct TNode {
        std::vector<TNode*> child;
        bool is_end;

        TNode() {
            child.resize(26);
            is_end = false;
        }
    };

    TNode* root;

public:
    Trie() {
        root = new TNode();     
    }

    TNode* _toEnd(string& s) {
        TNode* cur = root;
        for (char c: s) {
            TNode* nn = cur->child[c-'a'];
            if (nn == nullptr) {
                return nullptr;
            }
            cur = nn;
        }
        return cur;
    }
    
    void insert(string word) {
        TNode* cur = root;
        for (char c: word) {
            if (cur->child[c-'a'] == nullptr) {
                cur->child[c-'a'] = new TNode();
            }
            cur = cur->child[c-'a'];
        }
        cur->is_end = true;
    }
    
    bool search(string word) {
        TNode* en = _toEnd(word);
        if (en == nullptr) {
            return false;
        }
        return en->is_end;
    }
    
    bool startsWith(string prefix) {
        TNode* en = _toEnd(prefix);
        return en ? true : false;
    }
};

File: python/questions/Trie.py
from typing import List, Optional

class Trie:

    class TNode:
        def __init__(self):
            self.child: List[Optional["Trie.TNode"]] = [None] * 26
            self.is_end: bool = False

    def __init__(self):
        self.root: "Trie.TNode" = self.TNode()

    def _toEnd(self, word: str) -> Optional[TNode]:
        cur: "Trie.TNode" = self.root
        for c in word:
            bias: int = ord(c) - ord('a')
            if cur.child[bias] is None:
                return None
            cur = cur.child[bias]
        return cur

    def insert(self, word: str) -> None:
        cur: "Trie.TNode" = self.root
        for c in word:
            bias: int = ord(c) - ord('a')
            if cur.child[bias] is None:
                cur.child[bias] = self.TNode()
            cur = cur.child[bias]
        cur.is_end = True

    def search(self, word: str) -> bool:
        n = self._toEnd(word)
        if n:
            return n.is_end
        else:
            return False

    def startsWith(self, prefix: str) -> bool:
        n = self._toEnd(prefix)
        return True if n else False
